src/analysis/vector/geometry_checker/qgsgeometrygapcheck.cpp

```diff
@@ -367,19 +367,39 @@ bool QgsGeometryGapCheck::mergeWithNeighbor( const QMap<QString, QgsFeaturePool
     return false;
   }
 
-  QgsSpatialIndex neighbourIndex( QgsSpatialIndex::Flag::FlagStoreFeatureGeometries );
-  neighbourIndex.addFeatures( neighbours );
 
   QgsPolyline snappedRing;
   QgsVertexIterator iterator = errGeometry->vertices();
   while ( iterator.hasNext() )
   {
     QgsPoint pt = iterator.next();
     QgsVertexId id;
-    QgsGeometry closestGeom = neighbourIndex.geometry( neighbourIndex.nearestNeighbor( QgsPointXY( pt ) ).first() );
```

Comment:
Dropping the index is going to have large impacts on the speed of this test. Is there any way you could retain the index but make it instead store the boundaries of polygons instead, so that the nearest neighbour test will consider the distance from rings?

Reply:
For robust results, I think we'd have to build the spatial index on vertices, in which case I dont think there's a performance gain in using an index ? (it still somehow needs to iterate through all vertices).

The issue is that if we are checking a point that lies exactly on an other polygon's edge, the distance to the ring may very well be 0, that doesn't mean this ring is the one with the closest vertex (this false assumption was the cause of the bug being fixed).

So we'd need to check not only against the nearest feature, but against N nearest features, probably within a search radius R, and then still search the closest vertex on these features... I'm not sure what to use for N and R, and think in the particular context of the geometry checker, where it's quite uncommon to have gaps between more than 3-4 features, it would end up making no real difference at the cost of some complex logic.

What do you think ?

At the very least, I replaced `distance` with `distanceSquared` to save one operation per vertex.